[Cocoa] Fall back to playing the largest video in the main frame if there's no actively controlled element

whsieh opened the pull request:
#### 1c989ca2a15285de8c1720fa2b1d0887e46b153f
<pre>
[Cocoa] Fall back to playing the largest video in the main frame if there&apos;s no actively controlled element
<a href="https://bugs.webkit.org/show_bug.cgi?id=276263">https://bugs.webkit.org/show_bug.cgi?id=276263</a>
<a href="https://rdar.apple.com/131185180">rdar://131185180</a>

Reviewed by Tim Horton.

Augment `playPredominantOrNowPlayingMediaSession` so that it additionally detects and plays the
largest video element in the viewport, in the absence of any active Now Playing session.

* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::playPredominantOrNowPlayingMediaSession):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::startPlayingPredominantVideo):

Begin playing the largest (by area) video in the main frame that isn&apos;t already playing. Use this as
a fallback, only when no other video is being controlled.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/VideoControlsManager.mm:
(-[VideoControlsManagerTestWebView waitForVideoToPlay]):
(-[VideoControlsManagerTestWebView waitForVideoToPlay:]):
(-[VideoControlsManagerTestWebView isVideoPaused:]):
(-[VideoControlsManagerTestWebView waitForVideoToPause]):
(TestWebKitAPI::TEST(VideoControlsManager, StartPlayingLargestVideoInViewport)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/large-videos-with-audio.html:

Add an API test to exercise the change — only the first video (out of the three in the page) should
begin playing, since it covers most of the visible viewport.

Canonical link: <a href="https://commits.webkit.org/280710@main">https://commits.webkit.org/280710@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4c0bfa6f3ce278da8a6a828594945a7cf645473b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57383 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36711 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9858 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61005 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7828 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59511 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44335 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8016 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46472 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5540 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59413 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34448 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49560 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27336 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31230 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6871 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6831 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7142 "Found 1 new test failure: media/video-transformed.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62684 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1296 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7234 "Found 1 new test failure: workers/wasm-hashset-many.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53729 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1302 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49594 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53816 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/12685 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1105 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8568 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32540 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33625 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34710 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33371 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5175,7 +5175,72 @@ void WebPage::videoControlsManagerDidChange()
+        if (area <= largestArea)
+            continue;
+
+        constexpr auto minimumAreaForLargestMediaElement = 40000;
```
is this 200x200? any hint about where it came from or if it should be based on the viewport size instead? (and if not, maybe still written as 200*200?

whsieh replied:
Yep, came from exactly 200x200 (which I arbitrarily chose as a cutoff 😅). It's probably a better idea to make it scale based on viewport, for the context where this codepath will be used. I'll change it to be something like "25% of the viewport area" instead.

Thanks for the review!

whsieh committed:
[Cocoa] Fall back to playing the largest video in the main frame if there's no actively controlled element
https://bugs.webkit.org/show_bug.cgi?id=276263
rdar://131185180

Reviewed by Tim Horton.

Augment `playPredominantOrNowPlayingMediaSession` so that it additionally detects and plays the
largest video element in the viewport, in the absence of any active Now Playing session.

* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::playPredominantOrNowPlayingMediaSession):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::startPlayingPredominantVideo):

Begin playing the largest (by area) video in the main frame that isn't already playing. Use this as
a fallback, only when no other video is being controlled.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/VideoControlsManager.mm:
(-[VideoControlsManagerTestWebView waitForVideoToPlay]):
(-[VideoControlsManagerTestWebView waitForVideoToPlay:]):
(-[VideoControlsManagerTestWebView isVideoPaused:]):
(-[VideoControlsManagerTestWebView waitForVideoToPause]):
(TestWebKitAPI::TEST(VideoControlsManager, StartPlayingLargestVideoInViewport)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/large-videos-with-audio.html:

Add an API test to exercise the change — only the first video (out of the three in the page) should
begin playing, since it covers most of the visible viewport.

Canonical link: https://commits.webkit.org/280710@main
